tools/tpm2_key_import.c

```diff
@@ -0,0 +1,578 @@
+    TPM2B_DATA enc_sensitive_key;
+    TPM2B_MAX_BUFFER encrypted_inner_integrity;
+    TPM2B_MAX_BUFFER encrypted_duplicate_sensitive;
+    //SAPI and TCTI context
+    TSS2_SYS_CONTEXT *sapi_context;
+};
+
+int ssl_RSA_set0_key(RSA *r, BIGNUM *n, BIGNUM *e, BIGNUM *d) {
+
+    if ((r->n == NULL && n == NULL) || (r->e == NULL && e == NULL))
+        return 0;
+    if (n != NULL) {
+        BN_free(r->n);
+        r->n = n;
+    }
+    if (e != NULL) {
+        BN_free(r->e);
+        r->e = e;
+    }
+    if (d != NULL) {
+        BN_free(r->d);
+        r->d = d;
+    }
+    return 1;
+}
+
+bool encrypt_seed_with_tpm2_rsa_public_key(tpm_import_ctx *ctx) {
+    //Public Modulus
+    FILE *fp = fopen(ctx->parent_key_public_file, "rb");
+    if (fp == NULL) {
+        LOG_ERR("Failed accessing parent key public file.");
+        return false;
+    }
+    if (fseek(fp, 102, SEEK_SET) != 0) {
+        LOG_ERR("Expected parent key public data file size failure");
+        return false;
+    }
+    unsigned char pub_modulus[MAX_RSA_KEY_BYTES] = { 0 };
+    int ret = fread(pub_modulus, 1, MAX_RSA_KEY_BYTES, fp);
+    if (ret != MAX_RSA_KEY_BYTES) {
+        LOG_ERR("Failed reading public modulus from parent key public file");
+        return false;
+    }
+    fclose(fp);
+    RSA *rsa = NULL;
+    unsigned char encoded[MAX_RSA_KEY_BYTES];
+    unsigned char label[10] = { 'D', 'U', 'P', 'L', 'I', 'C', 'A', 'T', 'E', 0 };
+    int return_code = RSA_padding_add_PKCS1_OAEP_mgf1(encoded,
+            MAX_RSA_KEY_BYTES, ctx->protection_seed_data, 32, label, 10,
+            EVP_sha256(), NULL);
+    if (return_code != 1) {
+        printf("Failed RSA_padding_add_PKCS1_OAEP_mgf1\n");
+        return false;
+    }
+    BIGNUM* bne = BN_new();
+    return_code = BN_set_word(bne, RSA_F4);
+    if (return_code != 1) {
+        printf("BN_set_word failed\n");
+        return 1;
+    }
+    rsa = RSA_new();
+    return_code = RSA_generate_key_ex(rsa, 2048, bne, NULL);
+    if (return_code != 1) {
+        printf("RSA_generate_key_ex failed\n");
+        return 1;
+    }
+    BIGNUM *n = BN_bin2bn(pub_modulus, MAX_RSA_KEY_BYTES, NULL);
+    ssl_RSA_set0_key(rsa, n, NULL, NULL);
+    if (n == NULL) {
+        printf("Failed RSA_set0_key\n");
+        return 1;
+    }
+    // Encrypting
+    return_code = RSA_public_encrypt(MAX_RSA_KEY_BYTES, encoded,
+            ctx->encrypted_protection_seed_data, rsa, RSA_NO_PADDING);
+    if (return_code < 0) {
+        printf("Failed RSA_public_encrypt\n");
+    }
+    RSA_free(rsa);
+    BN_free(bne);
+    return true;
+}
+
+void aes_128_cfb_encrypt_buffers(uint8_t *buffer1, uint16_t buffer1_size,
+        uint8_t *buffer2, uint16_t buffer2_size, uint8_t *encryption_key,
+        uint8_t *encrypted_data) {
+    //AES encryption
+    uint8_t to_encrypt_buffer[max_buffer_size];
+    memcpy(to_encrypt_buffer, buffer1, buffer1_size);
+    memcpy(to_encrypt_buffer + buffer1_size, buffer2, buffer2_size);
```

Comment:
Do we really need this? Is it possible to overflow max_buffer_size?
Couldn't you just take an array of pointers sizes and encrypt them? This avoids the memcpy.

Maybe take an array of TPM2B pointers this way you have a structure with size, bytes.